proposed/verifiable-ref-readonly.md

```diff
@@ -0,0 +1,111 @@
+4) Additional special cases where use of a variable is considered readonly: 
+  (mostly affects codegeneration, does not have direct effect on semantics of the language):
+	- receiver of an instance member of a `readonly` struct
+	- receiver of a value type member that is inherited from a base class (`Enum`, `ValueType`, `Object`...)
+	- receiver of an instance member of `Nullable<T>`
+	- Controlled-mutability references that are obtained via unboxing or array indexing with `readonly.` prefix can be used as readonly references.
+
+NOTE: C# rules operate with high level language entities such as scopes, expressions, local variables, etc., that do not exist at the level of IL or do not map one-to-one.
+The following is the attempt to introduce/adjust verification rules to allow validation of the safety of the readonly references at IL level.
+
+# Verification of readonly references. #
+
+## Readonly managed pointer. ##
+
+CLI needs to add a notion of another transient type - readonly managed pointer. 
+*(need to adjust “III.1.8.1.1 Verification algorithm” and in “I.8.7 Assignment compatibility” accordingly)*
+
+NOTE: there are similarities with already existing concept of controlled mutability managed references, however `readonly` is a stronger property than controlled mutability. I.E. a controlled mutability reference is accepted anywhere where readonly reference could be used, but not the other way around.
+
+`readonly references` ⊂ `controlled mutability references` ⊂ `ordinary references`
+
+Readonly managed pointer is a transient type used for verification purposes. When certain conditions are met, managed pointers can be classified as readonly. 
+
+Unlike other managed pointer types, readonly managed pointers can be used:
+- as operands of `RET` instruction only if containing method returns _readonly_ references
+- as an operand of `LDFLD`, `LDIND`, `LDOBJ` and a source operand of `CPOBJ` 
+- as an operand of `LDFLDA` - the resulting pointer is considered _readonly_
+- as operands of a method call if corresponding parameter is _readonly_
+
+Parameters of a `CALL` or `CALLVIRT` instruction are considered readonly when they are
+- annotated as `in` parameters (see metadata encoding)
+- `this` parameters of `readonly` struct members (see metadata encoding)   
+- `this` parameters of struct members inherited from base types
+- `this` parameters of a member of `Nullable<T>`
+
+Readonly managed pointer may be a result of one of the following:
+-	`LDSFLDA`, `LDFLDA` when field is readonly 
+-	`LDARG` of an `in` parameter, 
+     note: arg0 inside an instance member other than .ctor of a readonly struct is considered an `in` parameter.
+-	`CALL`, `CALLVIRT` when the method returns a readonly reference
+
+## Merging stack states ## 
+*(add to III.1.8.1.3 Merging stack states)*
+
+Merging a readonly managed pointer with an managed pointer of an ordinary or controlled mutability kind results in a managed pointer of that kind.
```

Comment:
I believe this contradicts what's stated here about C# rules: https://github.com/dotnet/designs/pull/21/files#diff-f24baba7454a8bd449c0bce2f4e46336R15

Reply:
You are right. I adopted the part of this rule from the "returnable ref" proposal, but in this case the `readonly` should win over since it is the most constrained.

Will fix. Thanks!!